src/System.Management.Automation/engine/NativeCommandProcessor.cs

```diff
@@ -528,90 +541,346 @@ internal override void Complete()
+                                 stream == MinishellStream.Verbose ||
+                                 stream == MinishellStream.Warning)
+                        {
+                            //Convert to string
+                            try
+                            {
+                                obj = LanguagePrimitives.ConvertTo(obj, typeof(string), CultureInfo.InvariantCulture);
+                            }
+                            catch (PSInvalidCastException)
                             {
-                                startPosition.Y = 0;
+                                continue;
                             }
+                        }
+                        result.Add(new ProcessOutputObject(obj, stream));
+                    }
+                }
+            }
+            catch (XmlException originalException)
+            {
+                string template = NativeCP.CliXmlError;
+                string message = string.Format(
+                    null,
+                    template,
+                    isOutput ? MinishellStream.Output : MinishellStream.Error,
+                    Path,
+                    originalException.Message);
+                XmlException newException = new XmlException(
+                    message,
+                    originalException);
+
+                ErrorRecord error = new ErrorRecord(
+                    newException,
+                    "ProcessStreamReader_CliXmlError",
+                    ErrorCategory.SyntaxError,
+                    Path);
+                result.Add(new ProcessOutputObject(error, MinishellStream.Error));
+            }
 
-                            Host.BufferCell[,] bufferContents = this.Command.Context.EngineHostInterface.UI.RawUI.GetBufferContents(
-                                new Host.Rectangle(startPosition, endPosition));
+            return result;
+        }
 
-                            StringBuilder lineContents = new StringBuilder();
-                            StringBuilder bufferText = new StringBuilder();
+        private void InitOutputQueue()
+        {
+            //if output is redirected, start reading output of process in queue.
+            if (_nativeProcess.StartInfo.RedirectStandardOutput || _nativeProcess.StartInfo.RedirectStandardError)
+            {
+                lock (_sync)
+                {
+                    if (!_stopped)
+                    {
+                        _nativeProcessOutputQueue = new ConcurrentQueue<ProcessOutputObject>();
 
-                            for (int row = 0; row < bufferContents.GetLength(0); row++)
+                        if (_nativeProcess.StartInfo.RedirectStandardOutput)
+                        {
+                            bool isFirstOutput = true;
+                            bool isXmlCliOutput = false;
+                            _nativeProcess.OutputDataReceived += new DataReceivedEventHandler((sender, e) =>
```

Comment:
Can we encapsulate the event handlers and related variables/methods/objects into a class? I'm thinking about something like the following:
```CSharp
...
var handler = new ProcessOutputHandler(true, true);
_nativeProcess.OutputDataReceived += new DataReceivedEventHandler(handler.OutputHandler);
_nativeProcess.ErrorDataReceived += new DataReceivedEventHandler(handler.ErrorHandler);
...

public class ProcessOutputHandler
{
    public ProcessOutputHandler (bool redirectOutput, bool redirectError)
    {
        Debug.Assert(redirectOutput || redirectError, "Caller should redirect at least one stream");
        _blockingQueue = new BlockingCollection<Tuple<string, int>>(new ConcurrentQueue<Tuple<string, int>>());

        if (redirectOutput) { _refCount ++; }
        if (redirectError) { _refCount ++; }
    }

    private int _refCount = 0;
    private bool _isFirstOutput = true, _isFirstError = true;
    private bool _isOutputXml = false, _isErrorXml = false;
    private BlockingCollection<Tuple<string, int>> _blockingQueue;

    public BlockingCollection<Tuple<string, int>> BlockingQueue
    {
        get { return _blockingQueue; }
    }

    private List<ProcessOutputObject> DeserializeCliXmlObject(string xml, bool isOutput)
    {
        // ...
    }

    public void OutputHandler(object sendingProcess, DataReceivedEventArgs outLine)
    {
        if (outLine.Data == null)
        {
            // StandardOutput is closed
            if (Interlocked.Decrement(ref _refCount) == 0)
            {
                _blockingQueue.CompleteAdding();
            }
        }
        else 
        {
            // Process output and add to _blockingQueue ...
        }
    }

    public void ErrorHandler(object sendingProcess, DataReceivedEventArgs errorLine)
    {
        if (errorLine.Data == null)
        {
            // ErrorOutput is closed
            if (Interlocked.Decrement(ref _refCount) == 0)
            {
                _blockingQueue.CompleteAdding();
            }
        }
        else
        {
            // Process error and add to _blockingQueue ...
        }
    }
}
```